### PR TITLE
Multiple task ratings

### PR DESCRIPTION
**Pull request by IidaKainu**
Routes for adding/patching multiple ratings with one request. Either all or no ratings get added/patched.

In task rating modal user can only either patch or add ratings. (all changes need to be done in one request and it would not be good practice to make post and patch queries in the same route)

### CODE REVIEW

File: server/src/api/taskratings/taskratings.routes.ts [hunk 1]
```diff
@@ -16,20 +18,30 @@ taskratingRouter.get(
   requirePermission(Permission.TaskRatingRead),
   getTaskratings,
 );
+taskratingRouter.post(
+  '/taskrating',
```
Comment (juhoeerola) [inline server/src/api/taskratings/taskratings.routes.ts:22]: We could remove the singular versions of these routes. Or are these actually used somewhere?

File: server/src/api/taskratings/taskratings.controller.ts [hunk 2]
```diff
@@ -10,23 +10,74 @@ import Customer from '../customer/customer.model';
...
+
+  if (errors.some(({ err }: any) => err === 403))
+    return void (ctx.status = 403);
+
+  const newRatings = await Taskrating.transaction(
+    async (trx) =>
+      await Promise.all(
+        ratings.map(
+          async ({ dimension, value, comment, forCustomer }: any) =>
+            await Task.relatedQuery('ratings', trx)
```
Comment (juhoeerola) [inline server/src/api/taskratings/taskratings.controller.ts:62]:
```suggestion
   (trx) =>
     Promise.all(
        ratings.map(
         ({ dimension, value, comment, forCustomer }: any) =>
           Task.relatedQuery('ratings', trx)
```

The `async () => await ...` can be simplified to `() => ...` which returns equivalent promise.

File: server/src/api/taskratings/taskratings.controller.ts [hunk 3]
```diff
@@ -83,9 +129,58 @@ export const deleteTaskratings: RouteHandlerFnc = async (ctx) => {
...
+    return void (ctx.status = 404);
+  if (foundRatings.some(({ err }: any) => err === 403))
+    return void (ctx.status = 403);
+
+  const updated = await Taskrating.transaction(
+    async (trx) =>
+      await Promise.all(
+        foundRatings.map(
+          async ({ rating, value, comment }: any) =>
+            await rating.$query(trx).patchAndFetch({ value, comment }),
```
Comment (juhoeerola) [inline server/src/api/taskratings/taskratings.controller.ts:175]:
```suggestion
    (trx) =>
     Promise.all(
        foundRatings.map(
         ({ rating, value, comment }: any) =>
           rating.$query(trx).patchAndFetch({ value, comment }),
```

same here

File: server/src/api/taskratings/taskratings.controller.ts [hunk 4]
```diff
@@ -10,23 +10,74 @@ import Customer from '../customer/customer.model';
...
+        const found = await Customer.relatedQuery('representatives')
+          .for(rating.forCustomer)
+          .findById(userId);
+        if (!found) return { err: 403 };
+      }
+      return {};
+    }),
+  );
+
+  if (errors.some(({ err }: any) => err === 403))
```
Comment (juhoeerola) [inline server/src/api/taskratings/taskratings.controller.ts:54]:
```suggestion
  if (errors.some(({ err }) => err === 403))
```

The type inference is fine for this (`number | undefined`), same for two similar ones from line 165.

File: frontend/src/redux/roadmaps/reducers.ts [hunk 5]
```diff
@@ -165,43 +165,43 @@ export const DELETE_TASK_FULFILLED = (
...
-    parentTask.ratings.push(action.payload);
-  }
+  action.payload.forEach((taskRating) => {
+    let parentTask: Task | undefined;
+    state.roadmaps!.forEach((roadmap) => {
+      if (parentTask !== undefined) return;
+      parentTask = roadmap.tasks.find(
+        (task) => task.id === taskRating.parentTask,
+      );
+    });
```
Comment (juhoeerola) [inline frontend/src/redux/roadmaps/reducers.ts:181]: This is repeated three times, and sufficiently non-trivial to benefit from extracting out. For example:

```ts
const findTask = (taskId: number, roadmaps: Roadmap[]) =>
  roadmaps.reduce<Task | undefined>(
    (found, roadmap) =>
      found ?? roadmap.tasks.find((task) => task.id === taskId),
    undefined,
  );
```

```suggestion
    const parentTask = findTask(taskRating.parentTask, state.roadmaps!);
```

File: frontend/src/components/modals/RateTaskModal.tsx [hunk 6]
```diff
@@ -153,26 +156,33 @@ export const RateTaskModal: Modal<ModalTypes.RATE_TASK_MODAL> = ({
...
 
     setIsLoading(true);
-    const promises = businessValueRatings.flatMap(({ changed, ...rating }) =>
-      changed && rating.value !== undefined
-        ? [dispatch(roadmapsActions.addOrPatchTaskrating(rating))]
-        : [],
-    );
-    const results = await Promise.all(promises);
+    const changed = businessValueRatings.filter((rating) => rating.changed);
+    const error = edit ? await editRatings(changed) : await addRatings(changed);
```
Comment (juhoeerola) [inline frontend/src/components/modals/RateTaskModal.tsx:181]:
```suggestion
    const action = edit
      ? roadmapsActions.patchTaskratings
      : roadmapsActions.addTaskratings;
    const res = await dispatch(action({ ratings: changed, taskId }));
    const error = action.rejected.match(res) && res.payload?.message;
```

Could choose just the action conditionally, which makes the rest straight forward.

File: frontend/src/components/TaskTableUnrated.tsx [hunk 7]
```diff
@@ -91,21 +91,33 @@ const TableUnratedTaskRow: TableRow<Task> = ({ item: task, style }) => {
...
+    dispatch(
+      modalsActions.showModal({
+        modalType: ModalTypes.RATE_TASK_MODAL,
+        modalProps: {
+          taskId: task.id,
+          edit: false,
+        },
+      }),
+    );
+  };
```
Comment (juhoeerola) [inline frontend/src/components/TaskTableUnrated.tsx:119]:
```suggestion
  const openModal = (payload: ShowModalPayload) => (e: SyntheticEvent) => {
    e.preventDefault();
    e.stopPropagation();
    dispatch(modalsActions.showModal(payload));
  };

  const openRateModal = openModal({
    modalType: ModalTypes.RATE_TASK_MODAL,
    modalProps: {
      taskId: task.id,
      edit: false,
    },
  });
```

Could make the `openModal()` take `ShowModalPayload`. which would make it applicable for all three handlers.

**Review by juhoeerola — CHANGES_REQUESTED** [review]
Unless we use them, I'd remove the old singular routes.

Other than that, just few refactoring suggestions. The functionality seems good.

**Comment by IidaKainu** [comment]
Thanks again Juho for the insightful and neat refactoring suggestions. Looks like I need to brush up my knowledge on promises
Added fixes and rebased on master

**Comment by IidaKainu** [comment]
Forgot to edit tests

**Review by juhoeerola — APPROVED** [review]

**Pull request merged**